Kernel/FileSystem/InodeWatcher.cpp

```diff
@@ -1,98 +1,131 @@
 /*
  * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2021, the SerenityOS developers.
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
 #include <AK/Memory.h>
 #include <Kernel/FileSystem/Inode.h>
 #include <Kernel/FileSystem/InodeWatcher.h>
+#include <Kernel/FileSystem/WatchDescriptor.h>
+#include <Kernel/Process.h>
+#include <LibC/limits.h>
 
 namespace Kernel {
 
-NonnullRefPtr<InodeWatcher> InodeWatcher::create(Inode& inode)
+NonnullRefPtr<InodeWatcher> InodeWatcher::create()
 {
-    return adopt_ref(*new InodeWatcher(inode));
-}
-
-InodeWatcher::InodeWatcher(Inode& inode)
-    : m_inode(inode)
-{
-    inode.register_watcher({}, *this);
+    return adopt_ref(*new InodeWatcher());
 }
 
 InodeWatcher::~InodeWatcher()
 {
-    if (auto inode = m_inode.strong_ref())
-        inode->unregister_watcher({}, *this);
+    (void)close();
 }
 
-bool InodeWatcher::can_read(const FileDescription&, size_t) const
+KResult InodeWatcher::close()
 {
-    return !m_queue.is_empty() || !m_inode;
+    Locker locker(m_lock);
+
+    for (auto& entry : m_fd_to_watches) {
+        auto result = Process::current()->sys$close(entry.key);
```

Comment:
This just feels weird.

Reply:
Yeah, I couldn't find much of a better way to close the child watch descriptors. There doesn't seem to be a kernel-side method of closing fds on a process other than using `sys$close`, it seems.